### DISCUSSION TITLE
Best practice for complex brancing sequences / dialogue?

### DISCUSSION THREAD

**Discussion by Woodhouse3D** [ThePix/QuestJS#115] [Q&A]
For my game I'm planning to have lots of complex hyperlink-based branching dialogues and sequences. What are some best practices for coding them?

**Comment by KVonGit** [comment]
Hello.

Have you already seen this page?

https://github.com/ThePix/QuestJS/wiki/NPCs-%E2%80%90-Conversations-%E2%80%90-Dynamic-conversations

**Reply by Woodhouse3D** [reply]
yes, but I recently tried out the text adventure engine called Twine. (I'm currently testing it out given the evolving nature of my project) It uses passages that link to one another, kind of like the rooms system in Quest but more flexible. I think quest would really benefit from having some kind of system like this.

Side note: I'm also realizing I basically haven't played any of the old-school text advetures that Quest is trying to replicate... know any good ones?

**Reply by KVonGit** [reply]
### Quest Text Adventures
This is an easy-going parser that never purposefully frustrates you (great for beginners): 
https://textadventures.co.uk/games/view/z2yqtt1as0qxprpe1f5xag/the-bony-king-of-nowhere

This is another popular Quest parser, only this one uses the map and the panes:
https://textadventures.co.uk/games/view/sihv9g3rdki50mt9_1v4bq/xanadu-the-worlds-only-hope

---
### Pixie's QuestJS game
**The House on Highfield Lane**
https://textadventures.co.uk/games/view/idu7hlna7k2h24ib4ymmrg/the-house-on-highfield-lane

This is a good game. You'll probably get stuck in a couple of spots.

---
### Classic INFOCOM
Do you like to be frustrated? If not, don't play any INFOCOM games, haha. The main purpose of those old games is to frustrate the player. I love those games, but most people these days don't.

**The Hitchhiker's Guide to the the Galaxy** is my favorite:
https://eblong.com/infocom/gamefiles/hitchhiker-invclues-r31-s871119.z5 

NOTE: You can play any of the Z5 games I list here by copying their link after `https://iplayif.com/?story=`, like so:
https://iplayif.com/?story=https://eblong.com/infocom/gamefiles/hitchhiker-invclues-r31-s871119.z5

However, I advise playing them with Windows Frotz: https://github.com/DavidKinder/Windows-Frotz/releases/latest

There is also a GUI version here: https://www.bbc.co.uk/programmes/articles/1g84m0sXpnNCv84GpN2PLZG/the-game-30th-anniversary-edition

My second favorite INFOCOM game is **Leather Goddesses of Phobos**
https://eblong.com/infocom/gamefiles/leathergoddesses-r59-s860730.z3


Lot's of other folks like ZORK a lot:
https://eblong.com/infocom/gamefiles/zork1-invclues-r52-s871125.z5

Personally, I prefer Enchanter over Zork:
https://eblong.com/infocom/gamefiles/enchanter-r29-s860820.z3

Here's the complete list on that guy's site:
https://eblong.com/infocom/

Also, here are some helpful hints for a few of the INFOCOM games:
https://gallery.guetech.org/invisiclues.html

If you get into the INFOCOM games, you'll definitely like this:
https://archive.org/details/infocomcabinet

---
If you're looking for games that included conversations, though, those won't fit the bill.

You might be more into Emily Short's games:
https://emshort.blog/commercial-portfolio/

**Reply by KVonGit** [reply]
Oh.... Forgot...

This is a good Quest gamebook:
https://textadventures.co.uk/games/view/w207ce4zekubenmwgss5pa/victorian-detective

And this Squiffy game is all about conversation, and little more (it's also the number 1 top rated game on the site):
https://textadventures.co.uk/games/view/pnmno4m7guu6zyjropbleg/the-world-ends-today-i-think

**Comment by ThePix** [comment]
> "yes, but I recently tried out the text adventure engine called Twine. (I'm currently testing it out given the evolving nature of my project) It uses passages that link to one another, kind of like the rooms system in Quest but more flexible. I think quest would really benefit from having some kind of system like this."

If you can think of a way to do this, I would be very interested. The real issue is how you want to type the information - what format makes it easiest for the author. We are constrained to JavaScript, but other than that, we can be flexible - just create a function that will take your data and convert it into a set of QuestJS objects as usual.

Twine has a graphical interface that we cannot do, but behind that is something called Twee, and it would be (with some coding)  possible to convert from that. Give your NPC a "twee" attribute, which is a template literal - a multiline string with a backtick at either end. Then have some code that runs at start up that checks each object for a "twee" attribute and acts accordingly.

That said, you will probably need to extend Twee so the options only appear at the right time, and so options can impact the game world. And I found by the time I had taken account of that, I had arrived pretty much at where Dynamic Conversations is - which is why it is done like that.

**Comment by Woodhouse3D** [comment]
Ok, I've come up with a workaround- I teleport the player / or use a separate object to a new set of rooms. These rooms act as nodes in the dialogue tree, while naturally still allowing inventory use and other things. Then you use custom scope to keep the room you left in reach.
I'll see how it works later, and maybe it might be worthwhile to write up a tutorial on using the engine in this way.

**Comment by ThePix** [comment]
How about instead of teleporting the player object, having a different object that moves around the faux-rooms?
